data/model/namespacequota.py

```diff
@@ -0,0 +1,425 @@
+import json
+
+from peewee import fn, JOIN
+
+from data.database import (
+    Repository,
+    QuotaLimits,
+    UserOrganizationQuota,
+    QuotaType,
+    Manifest,
+    get_epoch_timestamp_ms,
+    Tag,
+    RepositorySize,
+    User,
+)
+from data import model
+from data.model import (
+    organization,
+    user,
+    InvalidUsernameException,
+    notification,
+    config,
+    InvalidSystemQuotaConfig,
+)
+
+
+def verify_namespace_quota(repository_ref):
+    model.repository.get_repository_size_and_cache(repository_ref._db_id)
+    namespace_size = get_namespace_size(repository_ref.namespace_name)
+    return check_limits(repository_ref.namespace_name, namespace_size)
+
+
+def verify_namespace_quota_force_cache(repository_ref):
+    force_cache_repo_size(repository_ref)
+    namespace_size = get_namespace_size(repository_ref.namespace_name)
+    return check_limits(repository_ref.namespace_name, namespace_size)
+
+
+def verify_namespace_quota_during_upload(repository_ref):
+    size = model.repository.get_size_during_upload(repository_ref._db_id)
+    namespace_size = get_namespace_size(repository_ref.namespace_name)
+    if namespace_size is None:
+        namespace_size = 0
+
+    return check_limits(repository_ref.namespace_name, size + namespace_size)
+
+
+def check_limits(namespace_name, size):
+    limits = get_namespace_limits(namespace_name)
+    limit_bytes = 0
+    severity_level = None
+    for limit in limits:
+        if size > limit["bytes_allowed"]:
+            if limit_bytes < limit["bytes_allowed"]:
+                limit_bytes = limit["bytes_allowed"]
+                severity_level = limit["type_id"]
```

Comment:
What if you have multiple limits defined, what is ordering based on when having multiple threshold.
For example, I have a limit set at 50% to send a warning, and another at 90% to error (does it make sense to have an error limit at something lower than 100%?),  and I'm at 95%, which one is this going to/should return?  

I don't think `get_namespace_limits` has a specific ordering.

Reply:
this loops through all limits and finds the largest limit that has been crossed and returns that one. So if your at 95% it would only return the reject. I cant think of only 1 real reason to set the threshold below 100% and that is to account for the system overshoot if it becomes a problem. Otherwise it is just a configuration allowed currently in 3.7. THis method has been confirmed by Daniel.